packages/dd-trace/src/appsec/iast/iast-plugin.js

```diff
@@ -0,0 +1,167 @@
+'use strict'
+
+const { channel } = require('../../../../diagnostics_channel')
+
+const iastLog = require('./iast-log')
+const Plugin = require('../../plugins/plugin')
+const appsecTelemetry = require('../telemetry')
+const { getInstrumentedMetric, getExecutedMetric, MetricTag } = require('./iast-metric')
+const { storage } = require('../../../../datadog-core')
+const { getIastContext } = require('./iast-context')
+const instrumentations = require('../../../../datadog-instrumentations/src/helpers/instrumentations')
+
+/**
+ * Used by vulnerability sources and sinks to subscribe diagnostic channel events
+ * and indicate what kind of metrics the subscription provides
+ * - moduleName is used identify when a module is loaded and
+ *    to increment the INSTRUMENTED_[SINK|SOURCE] metric when it occurs
+ * - channelName is the channel used by the hook to publish execution events
+ * - tag indicates the name of the metric: taint-tracking/source-types for Sources and analyzers type for Sinks
+ * - metricTag can be only SOURCE_TYPE (Source) or VULNERABILITY_TYPE (Sink)
+ */
+class IastPluginSubscription {
+  constructor (moduleName, channelName, tag, metricTag = MetricTag.VULNERABILITY_TYPE) {
+    this.moduleName = moduleName
+    this.channelName = channelName
+    this.tag = tag
+    this.metricTag = metricTag
+    this.executedMetric = getExecutedMetric(this.metricTag)
+    this.instrumentedMetric = getInstrumentedMetric(this.metricTag)
+  }
+
+  increaseInstrumented () {
+    this.instrumentedMetric.increase(this.tag)
+  }
+
+  increaseExecuted (iastContext) {
+    this.executedMetric.increase(this.tag, iastContext)
+  }
+}
+
+class IastPlugin extends Plugin {
+  constructor () {
+    super()
+    this.configured = false
+    this.pluginSubs = []
+  }
+
+  _wrapHandler (handler) {
+    return (message, name) => {
+      try {
+        handler(message, name)
+      } catch (e) {
+        iastLog.errorAndPublish(e)
+      }
+    }
+  }
+
+  _getTelemetryHandler (iastSub) {
+    return () => {
+      try {
+        const iastContext = getIastContext(storage.getStore())
+        iastSub.increaseExecuted(iastContext)
+      } catch (e) {
+        iastLog.errorAndPublish(e)
+      }
+    }
+  }
+
+  addSub (iastSub, handler) {
+    if (typeof iastSub === 'string') {
+      super.addSub(iastSub, this._wrapHandler(handler))
+    } else {
+      iastSub = this._getAndRegisterSubscription(iastSub)
+      if (iastSub) {
+        super.addSub(iastSub.channelName, this._wrapHandler(handler))
+
+        if (appsecTelemetry.isEnabled()) {
+          super.addSub(iastSub.channelName, this._getTelemetryHandler(iastSub))
+        }
+      }
+    }
+  }
+
+  onConfigure () {}
+
+  configure (config) {
+    if (!this.configured) {
```

Comment:
I think that if config if false, you don't want to call to `this.onConfigure()`, this is called from `iast/analyzers/index.js` on disable.
And if config is false, do you want to set `this.configured` as true?

Reply:
IastPlugins use `onConfigure` method to subscribe to certain channels so `this.configured` is a flag in order to do that susbscriptions only once. 
But yes, it can be more efficient if the IastPlugin doesn't subscribe if it is not enabled.